nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,439 @@
+            value: &impl serde::Serialize,
+        ) -> Result<steno::Node, super::SagaInitError> {
+            let value = serde_json::to_value(value).map_err(|e| {
+                SagaInitError::SerializeError(name.to_string(), e)
+            })?;
+            Ok(Node::constant(name, value))
+        }
+
+        builder.append(Node::action(
+            INSTANCE_LOCK_ID,
+            "GenerateInstanceLockId",
+            ACTION_GENERATE_ID.as_ref(),
+        ));
+        builder.append(become_updater_action());
+
+        // TODO(eliza): perhaps the start saga could determine whether it even
+        // needs to create a "real" saga based on whether either of the relevant
+        // state changes have occurred?
+
+        // If the active VMM's state has changed to "destroyed", then clean up
+        // after it.
+        if let Some(ref active_vmm) = params.state.active_vmm {
+            // If the active VMM is `Destroyed`, schedule the active VMM
+            // destroyed subsaga.
+            if active_vmm.runtime.state == VmmState::Destroyed {
+                builder.append(const_node(
+                    DESTROYED_VMM_ID,
+                    &PropolisUuid::from_untyped_uuid(active_vmm.id),
+                )?);
+                builder.append(destroyed_release_sled_resources_action());
+                builder.append(destroyed_release_virtual_provisioning_action());
+                builder.append(destroyed_unassign_oximeter_producer_action());
+                builder.append(destroyed_update_network_config_action());
+                builder.append(destroyed_update_instance_action());
+                builder.append(destroyed_mark_vmm_deleted_action());
+            }
+        }
+
+        // Next, determine what to do with the migration. A migration update
+        // saga needs to be scheduled if (and only if) the instance's migration
+        // ID currently points to a migration. The `instance_fetch_all` query
+        // will only return a migration if it is the instance's currently active
+        // migration, so if we have one here, that means that there's a
+        // migration.
+        if let Some(ref migration) = params.state.migration {
+            // If either side of the migration reports a terminal state, update
+            // the instance to reflect that.
+            if migration.is_terminal() {
+                builder.append(const_node(MIGRATION, migration)?);
+                // TODO(eliza): perhaps we could determine the final state in
+                // `make_saga_dag` and push a constant node for it, and then
+                // only have one `update_instance` action that's run regardless
+                // of which path through the saga we build...
+                builder.append(migration_update_instance_action());
+                builder.append(migration_update_network_config_action());
+            }
+        }
+
+        builder.append(unlock_instance_action());
```

Comment:
Because this is distinct from `destroyed_update_instance_action`, I think it might be possible to have a race where

- the destroyed-VMM action moves the instance back to `NoVmm`
- someone comes in and deletes the instance (legal, since there's no VMM)
- `instance_updater_unlock` observes that there's no record to unlock (since `time_deleted` is now set)

This looks like it will cause the update to unwind.

I think the simplest fix may be to drop the lock and apply any instance state updates in a single query. (This might be another reason to compute the "final" instance record and any attendant actions all in one go.)

Reply:
Yeah, dropping the lock and writing back the instance record in the same step also has the advantage of fewer total saga nodes. I'll take a crack at that.